File: runme/mcgen/local/datan/datsrc/minsim.f
      SUBROUTINE MINSIM(X0,N,NRED,LIST,USERFN,FMIN,EPSILN,NSTEP,X)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER(MAXN=50)
      DIMENSION X0(N),X(NRED,NRED+1)
      DIMENSION Y(MAXN+1),XEXP(MAXN),XR(MAXN),XR2(MAXN),
     +XPRIME(MAXN),X1(MAXN),X2(MAXN),XMX(MAXN),XMN(MAXN)
      DIMENSION LIST(N)
      PARAMETER (EPSDEF=1.D-25,TT=1.D-30,MAXSTP=1000)
      PARAMETER (ALPHA=1.D0,BETA=.5D0,GAMMA=2.D0)
      PARAMETER (ONE=1.D0,HALF=.5D0,ZERO=0.D0)
      EXTERNAL USERFN
C initialization
      EPS=EPSILN
      IF(EPS.LE.ZERO) EPS=EPSDEF
      NST=NSTEP
      IF(NST.LE.0) NST=MAXSTP
      M=NRED+1
C construct initial simplex and compute function at its corners
      Y(1)=USERFN(X0,N)
      CALL MTXGSV(X0,X1,N,NRED,LIST)
      CALL MTXPCL(X,X1,NRED,M,1)
      D=FMIN
      IF(D.LE.ZERO) D=1.D-5
      IRED=0
      DO 10 I=1,N
        IF(LIST(I).EQ.1) THEN
          IRED=IRED+1
          CALL MTXCPV(X1,X2,NRED)
          X2(IRED)=X2(IRED)+D
          CALL MTXPCL(X,X2,NRED,M,IRED+1)
          CALL MTXCPV(X0,XEXP,N)
          CALL MTXPSV(XEXP,X2,N,NRED,LIST)
          Y(IRED+1)=USERFN(XEXP,N)
        END IF
   10 CONTINUE
      ISTEP=0
C begin iteration
   20 ISTEP=ISTEP+1
C compute indices of corner with largest (MX), next to largest (NE)
C and smallest (MN) function value
      MN=1
      IF(Y(1).GT.Y(2)) THEN
        MX=1
        NE=2
      ELSE
        MX=2
        NE=1
      END IF
      DO 30 I=1,NRED+1
        IF(Y(I).LT.Y(MN)) MN=I
        IF(Y(I).GT.Y(MX)) THEN
          NE=MX
          MX=I
        ELSE IF (Y(I).GT.Y(NE)) THEN
          IF(I.NE.MX) NE=I
        END IF
   30 CONTINUE
C test for break-off condition
      IF(ABS(Y(MX)-Y(MN)).LT.EPS*ABS(Y(MN))+TT .OR. ISTEP.GT.NST)
     +GO TO 60
C compute XPRIME, i.e. point in center of gravity of hyperplane
C opposite to point with largest function value
      CALL MTXGCL(X,XMX,NRED,M,MX)
      CALL MTXGCL(X,XMN,NRED,M,MN)
      CALL MTXZRV(XPRIME,NRED)
      DO 40 I=1,NRED+1
        IF(I.NE.MX) THEN
          CALL MTXGCL(X,X1,NRED,M,I)
          CALL MTXADV(XPRIME,X1,XPRIME,NRED)
        END IF
   40 CONTINUE
      F=ONE/DBLE(NRED)
      CALL MTXMSV(XPRIME,XPRIME,F,NRED)
C construct points by reflection (XR) and extension (XR2)
      CALL MTXMSV(XPRIME,X1,ONE+ALPHA,NRED)
      CALL MTXMSV(XMX,X2,ALPHA,NRED)
      CALL MTXSBV(X1,X2,XR,NRED)
      CALL MTXCPV(X0,XEXP,N)
      CALL MTXPSV(XEXP,XR,N,NRED,LIST)
      YR=USERFN(XEXP,N)
      IF(YR.LE.Y(MN)) THEN
        CALL MTXMSV(XR,X1,GAMMA,NRED)
        CALL MTXMSV(XPRIME,X2,ONE-GAMMA,NRED)
        CALL MTXADV(X1,X2,XR2,NRED)
        CALL MTXCPV(X0,XEXP,N)
        CALL MTXPSV(XEXP,XR2,N,NRED,LIST)
        YR2=USERFN(XEXP,N)
        IF(YR2.LT.Y(MN)) THEN
C perform extension
          CALL MTXPCL(X,XR2,NRED,M,MX)
          Y(MX)=YR2
        ELSE
C perform reflection
          CALL MTXPCL(X,XR,NRED,M,MX)
          Y(MX)=YR
        END IF
      ELSE IF(YR.GE.Y(NE)) THEN
        IF(YR.LT.Y(MX)) THEN
C perform reflection
          CALL MTXPCL(X,XR,NRED,M,MX)
          Y(MX)=YR
        END IF
        CALL MTXMSV(XMX,X1,BETA,NRED)
        CALL MTXMSV(XPRIME,X2,ONE-BETA,NRED)
        CALL MTXADV(X1,X2,XR2,NRED)
        CALL MTXCPV(X0,XEXP,N)
        CALL MTXPSV(XEXP,XR2,N,NRED,LIST)
        YR2=USERFN(XEXP,N)
        IF(YR2.LT.Y(MX)) THEN
C perform compression
          CALL MTXPCL(X,XR2,NRED,M,MX)
          Y(MX)=YR2
        ELSE
C perform contraction
          DO 50 I=1,NRED+1
            IF(I.NE.MN) THEN
              CALL MTXGCL(X,X1,NRED,M,I)
              CALL MTXADV(X1,XMN,X2,NRED)
              CALL MTXMSV(X2,X2,HALF,NRED)
              CALL MTXPCL(X,X2,NRED,M,I)
              CALL MTXCPV(X0,XEXP,N)
              CALL MTXPSV(XEXP,X2,N,NRED,LIST)
              Y(I)=USERFN(XEXP,N)
            END IF
   50     CONTINUE
        END IF
      ELSE
C perform reflection
        CALL MTXPCL(X,XR,NRED,M,MX)
        Y(MX)=YR
      END IF
      GO TO 20
C Fill output arguments
   60 ISTEP=ISTEP-1
      FMIN=Y(MN)
      CALL MTXGCL(X,X1,NRED,M,MN)
      CALL MTXPSV(X0,X1,N,NRED,LIST)
      IF(ISTEP.GT.NST) THEN
        NSTEP=-1
      ELSE
        NSTEP=ISTEP
      END IF
      END

File: runme/mcgen/local/datan/datsrc/minsim.c
#include <stdlib.h>
#include <limits.h>
#include "datsrc.h"

void LIBFUNC minsim_(double P_T *x0, integer const P_T *n,
             integer const P_T *nred, integer const P_T *list,
             double (CBFUNC P_T *userfn)(double const P_T *,
             integer const P_T *), double P_T *fmin,
             double const P_T *epsiln, integer P_T *nstep, double P_T *x)
{
    /* Table of constant values */

    integer c__1 = 1;
    double c_d2 = 2., c_d1 = 1., c_dm1 = -1., c_dp5 = .5;

    /* System generated locals */
    integer x_dim1, x_offset, i__1, i__2;
    double d__1, d__2;

    /* Local variables */
    integer i, m, ired, istep, ne, mn, mx, nst;
    double d, f, yr, yr2, eps;
    /*double xexp[50], xprime[50], xr[50], xr2[50], xmx[50], xmn[50],
           x1[50], x2[50], y[51];*/
    double *xexp, *xprime, *xr, *xr2, *xmx, *xmn, *x1, *x2, *y;
    unsigned long amem;

    if ((amem = sizeof(double) * (*n + 8 * *nred + 1)) > MAXALLOC) {
        MEMERROR("minsim_ (too large)");
        goto f_0;
    }
    xexp = (double *)MEMALLOC(amem);
    if (!xexp) {
        MEMERROR("minsim_");
        goto f_0;
    }
    xprime = xexp + *n; xr = xprime + *nred; xr2 = xr + *nred;
    xmx = xr2 + *nred; xmn = xmx + *nred; x1 = xmn + *nred; x2 = x1 + *nred;
    y = x2 + *nred;

/* initialization */
    /* Parameter adjustments */
    --list;
    --x0;
    x_dim1 = *nred;
    x_offset = x_dim1 + 1;
    x -= x_offset;

    /* Function Body */
    eps = *epsiln;
    if (eps <= 0.) {
        eps = 1e-25;
    }
    nst = *nstep;
    if (nst <= 0) {
        nst = 1000;
    }
    m = *nred + 1;
/* construct initial simplex and compute function at its corners */
    y[0] = (*userfn)(&x0[1], n);
    mtxgsv_(&x0[1], x1, n, nred, &list[1]);
    mtxpcl_(&x[x_offset], x1, nred, &m, &c__1);
    d = *fmin;
    if (d <= 0.) {
        d = 1e-5;
    }
    ired = 0;
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
        if (list[i] == 1) {
            ++ired;
            mtxcpv_(x1, x2, nred);
            x2[ired - 1] += d;
            i__2 = ired + 1;
            mtxpcl_(&x[x_offset], x2, nred, &m, &i__2);
            mtxcpv_(&x0[1], xexp, n);
            mtxpsv_(xexp, x2, n, nred, &list[1]);
            y[ired] = (*userfn)(xexp, n);
        }
    }
    istep = 0;
/* begin iteration */
L20:
    ++istep;
/* compute indices of corner with largest (MX), next to largest (NE) */
/* and smallest (MN) function value */
    mn = 1;
    if (y[0] > y[1]) {
        mx = 1;
        ne = 2;
    } else {
        mx = 2;
        ne = 1;
    }
    i__1 = *nred + 1;
    for (i = 1; i <= i__1; ++i) {
        if (y[i - 1] < y[mn - 1]) {
            mn = i;
        }
        if (y[i - 1] > y[mx - 1]) {
            ne = mx;
            mx = i;
        } else if (y[i - 1] > y[ne - 1]) {
            if (i != mx) {
                ne = i;
            }
        }
    }
/* test for break-off condition */
    if ((d__1 = y[mx - 1] - y[mn - 1], abs(d__1)) < eps * (d__2 = y[mn - 1],
            abs(d__2)) + 1e-30 || istep > nst) {
        goto L60;
    }
/* compute XPRIME, i.e. point in center of gravity of hyperplane */
/* opposite to point with largest function value */
    mtxgcl_(&x[x_offset], xmx, nred, &m, &mx);
    mtxgcl_(&x[x_offset], xmn, nred, &m, &mn);
    mtxzrv_(xprime, nred);
    i__1 = *nred + 1;
    for (i = 1; i <= i__1; ++i) {
        if (i != mx) {
            mtxgcl_(&x[x_offset], x1, nred, &m, &i);
            mtxadv_(xprime, x1, xprime, nred);
        }
    }
    f = 1. / (double) (*nred);
    mtxmsv_(xprime, xprime, &f, nred);
/* construct points by reflection (XR) and extension (XR2) */
    mtxmsv_(xprime, x1, &c_d2, nred);
    mtxmsv_(xmx, x2, &c_d1, nred);
    mtxsbv_(x1, x2, xr, nred);
    mtxcpv_(&x0[1], xexp, n);
    mtxpsv_(xexp, xr, n, nred, &list[1]);
    yr = (*userfn)(xexp, n);
    if (yr <= y[mn - 1]) {
        mtxmsv_(xr, x1, &c_d2, nred);
        mtxmsv_(xprime, x2, &c_dm1, nred);
        mtxadv_(x1, x2, xr2, nred);
        mtxcpv_(&x0[1], xexp, n);
        mtxpsv_(xexp, xr2, n, nred, &list[1]);
        yr2 = (*userfn)(xexp, n);
        if (yr2 < y[mn - 1]) {
/* perform extension */
            mtxpcl_(&x[x_offset], xr2, nred, &m, &mx);
            y[mx - 1] = yr2;
        } else {
/* perform reflection */
            mtxpcl_(&x[x_offset], xr, nred, &m, &mx);
            y[mx - 1] = yr;
        }
    } else if (yr >= y[ne - 1]) {
        if (yr < y[mx - 1]) {
/* perform reflection */
            mtxpcl_(&x[x_offset], xr, nred, &m, &mx);
            y[mx - 1] = yr;
        }
        mtxmsv_(xmx, x1, &c_dp5, nred);
        mtxmsv_(xprime, x2, &c_dp5, nred);
        mtxadv_(x1, x2, xr2, nred);
        mtxcpv_(&x0[1], xexp, n);
        mtxpsv_(xexp, xr2, n, nred, &list[1]);
        yr2 = (*userfn)(xexp, n);
        if (yr2 < y[mx - 1]) {
/* perform compression */
            mtxpcl_(&x[x_offset], xr2, nred, &m, &mx);
            y[mx - 1] = yr2;
        } else {
/* perform contraction */
            i__1 = *nred + 1;
            for (i = 1; i <= i__1; ++i) {
                if (i != mn) {
                    mtxgcl_(&x[x_offset], x1, nred, &m, &i);
                    mtxadv_(x1, xmn, x2, nred);
                    mtxmsv_(x2, x2, &c_dp5, nred);
                    mtxpcl_(&x[x_offset], x2, nred, &m, &i);
                    mtxcpv_(&x0[1], xexp, n);
                    mtxpsv_(xexp, x2, n, nred, &list[1]);
                    y[i - 1] = (*userfn)(xexp, n);
                }
            }
        }
    } else {
/* perform reflection */
        mtxpcl_(&x[x_offset], xr, nred, &m, &mx);
        y[mx - 1] = yr;
    }
    goto L20;
/* Fill output arguments */
L60:
    --istep;
    *fmin = y[mn - 1];
    mtxgcl_(&x[x_offset], x1, nred, &m, &mn);
    mtxpsv_(&x0[1], x1, n, nred, &list[1]);
    if (istep > nst) {
        *nstep = -1;
    } else {
        *nstep = istep;
    }
    MEMFREE((void *)xexp);
f_0:
    return;
} /* minsim_ */

